public/js/src/enketo-webform.js

```diff
@@ -63,21 +95,41 @@ function _updateMaxSizeSetting( survey ) {
     return survey;
 }
 
+/**
+ * Wrap location access to detect/prevent navigation in tests.
+ */
+ const _location = {
+    get href() {
+        return location.href;
+    },
+    set href( href ) {
+        location.href = href;
+    },
+    reload: location.reload.bind( location ),
+};
+
+const LOAD_ERROR_CLASS = 'fail';
+const LOAD_ERROR_ENTRY_ADVICE = 'alert.loaderror.entryadvice';
+const LOAD_ERROR_UNKNOWN = 'error.unknown';
+
 function _showErrorOrAuthenticate( error ) {
     error = ( typeof error === 'string' ) ? new Error( error ) : error;
-    loader.classList.add( 'fail' );
+    loader.classList.add( LOAD_ERROR_CLASS );
 
     if ( error.status === 401 ) {
-        window.location.href = `${settings.loginUrl}?return_url=${encodeURIComponent( window.location.href )}`;
+        _location.href = `${settings.loginUrl}?return_url=${encodeURIComponent( _location.href )}`;
     } else {
         if ( !Array.isArray( error ) ) {
-            error = [ error.message  || t( 'error.unknown' ) ];
+            error = [ error.message  || t( LOAD_ERROR_UNKNOWN ) ];
```

Comment:
Just a single random comment so I won't forget. 

We used to have a build task called `grunt shell: translation` that updated the translation keys in locales/source/en/translation.json. It broke at some point and was disabled but I found it useful and wanted to bring it back eventually. https://github.com/i18next/i18next-parser

This parser doesn't support variables only string literals (as `t()` parameters). 

What do you think of the benefits of that (and manually updating translation keys) vs the benefits of exporting these for tests?

Reply:
Thanks for catching this!

I'm not sure I have an opinion either way about how translation keys are updated, I haven't really touched that at all so I'm not familiar. I didn't realize this depended on a parser at all. Is that necessary? It seems like it's doing the equivalent of `lodash.get`.

That said, addressing that here definitely feels out of scope, so my inclination is to revert to string literals and use the same string literals in the tests.